sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,731 @@
+  auto m2 = v2.format<float, 2, 2>();
+
+  // ...
+  m2 = m1.select<2, 2, 2, 4>(1, 2);  // v_size = 2, v_stride = 2,
+                                     // h_size = 2, h_stride = 4,
+                                     // v_offset = 1, h_offset = 2.
+```
+<p align="center">
+<img src="images/Matrix_2_2_2_4__1_2.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+```cpp
+  m1.select<4, 1, 4, 2>(0, 0) = 0.0f; // selected elements of m1
+                                      // are replaced with 0.0f
+```
+
+<p align="center">
+<img src="images/Matrix_4_1_4_2__0_0.svg" title="2D select example" width="700" height="200"/>
+</p>
+
+### Reduction functions
+
+Explicit SIMD provides the following reduction functions for simd objects.
+Compiler will produce optimal code sequence on the target device to apply the
+specified operation to all scalar elements in the input simd vector. Note that
+the order of element-wise operations is not guaranteed and the correctness of
+result should not depend on a particular computation order.
+- ```cm_sum(simd<T1, SZ> v)```: returns a scalar value of type **T0** that's
+equal to the sum of all data elements of type **T1** in simd vector **v** of
+length **SZ**.
+- ```cm_prod(simd<T1, SZ> v)```: returns a scalar value of type **T0** that's
+equal to the product of all data elements of type **T1** in simd vector **v** of
+length **SZ**.
+- ```cm_reduced_max(simd<T1, SZ> v)```: returns a scalar value of type **T0**
+that's equal to the maximum value of all data elements of type **T1** in simd
+vector **v** of length **SZ**.
+- ```cm_reduced_min(simd<T1, SZ> v)```: returns a scalar value of type **T0**
+that's equal to the minimum value of all data elements of type **T1** in simd
+vector **v** of length **SZ**.
+
+```cpp
+template <typename T0, typename T1, int SZ> T0 cm_sum(simd<T1, SZ> v);
+template <typename T0, typename T1, int SZ> T0 cm_prod(simd<T1, SZ> v);
+template <typename T0, typename T1, int SZ> T0 cm_reduced_max(simd<T1, SZ> v);
+template <typename T0, typename T1, int SZ> T0 cm_reduced_min(simd<T1, SZ> v);
+```
+
+### Memory access APIs
+
+Currently the variety of memory objects supported by the Explicit SIMD extension
+implementation is limited to the following:
+- USM pointers (aka 'flat address')
+- 1D global accessors
+- 2D image accessors
+
+The implementation further limits the set of memory access operations
+which a kernel can perform through those memory objects. Basically, all of them
+are special intrinsic APIs described below rather than standard SYCL accessor or
+pointer operations. Examples of unsupported features include
+`accessor::get_pointer()`, accessor's `operator []`, C/C++ dereference of an
+USM pointer, local accessors, 2D and 3D accessors.
+Those are temporary restrictions, to be dropped in future.
+
+See auto-generated documentation for the complete list of APIs here. (TBD)
+
+#### USM pointer-based memory access
+##### Flat-address gather/scatter
```

Comment:
This comment shouldn't block this PR, but I'd like us to think more about what we can do to make these interfaces more user-friendly.

Two things stand out:
1) A user providing cache hints has to provide a lot more template arguments than required.  Could we make this nicer by providing the hints as tag-type arguments?
    ```c++
   // This works
   float* p;
   simd<uint32_t, 16> offsets;
   auto result = flat_load(p, offsets);

   // Adding cache hints makes it much more complex
   float* p;
   simd<uint32_t, 16> offsets;
   auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);
    ```

2)  I find the existing names quite confusing.  I'd much prefer that the names were more obvious (e.g. `flat_load` could be called `gather` if that's all it does) and/or used overloading where appropriate to minimize the number of times a reader of code has to check documentation to figure out what a function is supposed to be doing.

Reply:
Makes sense to me. @kychendev, please comment.

Reply:
I'm open to change the name to make it more user friendly. The cache hints setting are mostly based on existing implementation. Please check with Gang who added these memory access APIs.

Reply:
I do not see why gather/scatter is better naming than load/store. Our current naming scheme has its own consistency.  I also do not see why tagged argument is any better than template

Reply:
The more explicit and obvious a name is, the better.  Anybody reading a function called "gather" or "gather load" knows that the function being called performed is a gather operation.  I wouldn't know that a "flat load" was a gather without reading the documentation.

A tagged argument is better in this case because it allows for more template arguments to be deduced.  Expanding on my previous example:

```c++
// Without cache hints, type and length can be deduced from offsets
float* p;
simd<uint32_t, 16> offsets;
auto result = flat_load(p, offsets);

// With cache hints as templates, verbosity increases significantly:
// - Providing any cache hint forces the user to specify the type and length
float* p;
simd<uint32_t, 16> offsets;
auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);

// With cache hints as tag types, verbosity is reduced:
// - Providing a cache hint does not prevent deduction of type and length
float* p;
simd <uint32_t, 16> offsets;
auto result = flat_load(p, offsets, CacheHint::Foo{});
```

Note also that the templated form prevents a developer from specifying an L3 hint without also explicitly specifying an L1 hint.  If `flat_load` accepted a list of hints, it might be possible to refactor the hints to specify them in any order, and it may be more extensible to future cache hints:
```c++
auto result = flat_load(p, offsets, l3_cache_hint_foo, ...);
```

Reply:
load/store means gather/scatter unless it is block_load/block_store

Reply:
I understand that, and the distinction is clear to somebody who has read the documentation.

I'm arguing that ideally we wouldn't be forcing developers to read documentation in order to make sense of what code does.

Reply:
Yeah, who can write code with reading documents? especially with modern C++.

Reply:
I didn't say _write_ code -- I said _read_ code.  Code is read more than it is written, and having good names for functions helps people who didn't write a particular piece of code to understand what it's supposed to do.  Having good names is helpful for training material, code review, onboarding new programmers, etc.

I think these names could be better, and I think we'll waste time explaining that `flat_load` is a gather operation every time we show somebody an example code using explicit SIMD types.  This isn't a hill I'm prepared to die on, so I'll leave it at that.

Reply:
@cmc-rep are you being facetious about reading documentation?  Descriptive names and interfaces are very important for readable/maintainable/accessible code. 

'flat' addresses are only meaningful on Gen, and the convention that load == gather is obviously a Gen-specific notion too, since we keep having to say 'oh this is a gather' over and over just in this conversation! 

Maybe it would be best to change the name of this extension back to DPC++-CM?

Reply:
Reading documentation is very dangerous when designing/writing write-only programs... :-)

Reply:
Also another area where it needless differs from `std::simd`. Non-gather load/store is called `copy_from` and `copy_to`. Why invent new names just to confuse people?

Reply:
> ```c++
> // Without cache hints, type and length can be deduced from offsets
> float* p;
> simd<uint32_t, 16> offsets;
> auto result = flat_load(p, offsets);
> auto result = flat_load<uint32_t, 16, 1, CacheHint::Foo, CacheHint::Bar>(p, offsets);
>```

An approach *à la* https://github.com/chriskohlhoff/propria from @chriskohlhoff would be to add a property to the pointer, such as

```c++
auto result = flat_load(p, offsets);
auto result = flat_load(decorate<CacheHint::Foo, CacheHint::Bar>(p), offsets);
```
The advantage is that you do not have to change all tour API and all the uses of this
decorated pointer will benefit from this.
`decorate` is to be bikeshed accordingly.

Reply:
added TODO to the headers